Automattic/jetpack
# E2E tests: merge github workflows

#### Changes proposed in this Pull Request:

There are currently 3 different Github workflows defined for e2e tests, and with this approach, a new one would have been created for each new e2e project for any new plugins.

This PR merges all these workflows into a single one, to add consistency and maintainability.

`.github/files/create-e2e-projects-matrix.sh` will evaluate workflow triggers, changed files, compare them with the `ci.targets` entries in the project's `package.json`, and decide which tests should run and the test arguments. The output is a JSON structure, containing an array to be used as the matrix for the e2e job. Entries in the matrix will run in parallel, and a single test report for all parallel runs will be created.
Slack notifications are sent individually, for each job.

This adds hard requirements for e2e projects:
* a set of scripts to exist in their `package.json`
    * `pnpm run build` - build all plugins that are required for the tests to successfully run
    * `pnpm run config:decrypt` - decrypt configuration
    * `pnpm run env:up` - starts and configures the Docker environment
    * `pnpm run tunnel:up` - creates a tunnel
    * `pnpm run tunnel:down`
    * `pnpm run test:run` - run the tests
    * `pnpm run slack`- calls the slack notification script

**Known issues (to be addressed in follow-ups):**
* `tools/e2e-commons` is not detected as changed by `list-changed.projects.sh`, so changes in this folder will not trigger the tests.
* better handling of unexpected `ci.targets` values in `package.json` (missing fields, string values instead of an array, etc.). Maybe `create-e2e-projects-matrix.sh` should be rewritten as a node script, that will make it easier to work with.
* add more flexibility when specifying test targets. Maybe use include/exclude patterns.

#### Jetpack product discussion
pd5faL-35-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* This one is hard to test end-to-end without merging first.
* For this PR, both Jetpack and Boost tests should run.
* Slack notifications should be sent for each of the 2 test groups/projects.
* A single test report for the PR, containing both Jetpack and Boost tests should be created.
* `.github/files/create-e2e-projects-matrix.sh` can be tested in isolation
